### PR TITLE
Enrich local data via the OpenAIRE Graph

### PR DESCRIPTION
**Pull request by LucaGiamminonni**
## References
* Related to https://github.com/DSpace/RestContract/pull/181
*  Related to https://github.com/DSpace/DSpace/pull/8184

## Description
The features included in this PR are the result of the OpenAIRE Call Innovation funded project "Enrich local data via the OpenAIRE Graph” awarded by 4Science (https://www.openaire.eu/open-call-winner-phase-1-4science). It provides a closer integration between DSpace and two OpenAIRE services, the Notification Broker and the OpenAIRE REST API.
Detailed documentation about the aims of the project, the implementation and the configuration options is available at https://4science.github.io/oaire-eld/#/
This PR regarding only the data correction section. The publication claim will be migrated with other PR.

## How to Use / Test

Video: https://github.com/DSpace/DSpace/assets/483997/6db3dd92-1e26-416c-b130-9f5677f7acaa

1. Start up frontend & backend
2. Import a valid JSON OpenAIRE Graph JSON file using the `import-openaire-events` Process (in the Processes Admin UI)
    * Sample file: https://github.com/DSpace/DSpace/files/13505259/test-nbevents-sample.-.Copie.json 
    * Structure of this JSON is defined at https://github.com/4Science/oaire-eld/blob/main/data-correction.md
    * "ENRICH" types defined at https://api.openaire.eu/vocabularies/dnet:topic_types
3. Assuming import is successful, this will create several Notifications on the `Admin menu -> Notifications -> Quality Assurance` page.  (The exact number/type of Notifications is defined in the JSON file you imported)
4. From that page, you can now use the "Actions" button to view each notification and decide whether to Accept/Ignore/Reject each.  
    * Accepting will apply the change.
    * Ignoring will delete the change without applying
    * Reject will delete the change without applying and send a notification back to the service.

### CODE REVIEW

File: src/app/admin/admin-notifications/admin-notifications-broker-source-page-component/admin-notifications-broker-source-page.component.ts
Comment (paulo-graca) [inline src/app/admin/admin-notifications/admin-notifications-broker-source-page-component/admin-notifications-broker-source-page.component.ts:7]: missing typescript docs

File: src/app/admin/admin-notifications/admin-notifications-broker-events-page/admin-notifications-broker-events-page.component.ts
Comment (paulo-graca) [inline src/app/admin/admin-notifications/admin-notifications-broker-events-page/admin-notifications-broker-events-page.component.ts:7]: missing typescript docs

**Review by paulo-graca — COMMENTED** [review]
Thank you @LucaGiamminonni! I've just noticed some missing typescript docs. They can be useful if someone tries to override this code to better understand what is the purpose.

File: src/app/notifications/broker/events/notifications-broker-events.component.ts
Comment (tdonohue) [inline src/app/notifications/broker/events/notifications-broker-events.component.ts:422]: This method has many hardcoded URL prefixes which sites may need to occasionally change.  I'm wondering if there's a way to move this logic to the backend where it could be more easily configurable?  Is there a reason why the prefix needs to be added here in the UI, instead of just making sure the full URL can be sent by the backend?

File: src/app/notifications/broker/events/notifications-broker-events.scomponent.scss
Comment (tdonohue) [inline src/app/notifications/broker/events/notifications-broker-events.scomponent.scss:1]: This file name has an extra `s` before `component`.  It says `.scomponent.scss`

File: src/app/notifications/broker/source/notifications-broker-source.component.ts
Comment (tdonohue) [inline src/app/notifications/broker/source/notifications-broker-source.component.ts:17]: Please add TypeDocs

**Review by tdonohue — COMMENTED** [review]
@LucaGiamminonni : I'm no longer able to test this as the backend doesn't compile & this PR has merge conflicts.  But, I did give this a code review today. Overall the code looks well commented & well designed. Obviously we'll need to rename many of these classes eventually to move from "notification broker" (NB) to "quality assurance" (QA) as decided in a dev meeting.  Beyond that, I just have a few other inline comments.

Once this & the backend are ready for testing again, I'll gladly give this a test so that I can comment on the UI, user experience, etc.

**Comment by atarix83** [comment]
@tdonohue  @paulo-graca  the PR is ready to be review. thanks

File: src/app/shared/mocks/notifications.mock.ts
Comment (tdonohue) [inline src/app/shared/mocks/notifications.mock.ts:83]: Silly thing, but I'd really like to make all URLs generic and avoid referencing old XMLUI.  There are several in this file which start with "http://dspace7.4science.it/xmlui/".  Could we change these all to start instead with either `https://demo.dspace.org` or a fake URL like `https://example.org`

File: src/app/suggestion-notifications/qa/events/quality-assurance-events.component.html
Comment (tdonohue) [reply src/app/suggestion-notifications/qa/events/quality-assurance-events.component.html:61]: Per the CodeQL alert on this line of code, this `<a>` tag should likely have `rel="noopener noreferrer"` added as attributes, since it has `target="_blank"`.

File: src/app/suggestion-notifications/qa/events/quality-assurance-events.component.html
Comment (tdonohue) [inline src/app/suggestion-notifications/qa/events/quality-assurance-events.component.html:87]: This link has a hardcoded URL back to OpenAIRE.  We may want to make this configurable at some point in case the URL changes

File: src/app/suggestion-notifications/qa/project-entry-import-modal/project-entry-import-modal.component.html
Comment (tdonohue) [inline src/app/suggestion-notifications/qa/project-entry-import-modal/project-entry-import-modal.component.html:12]: This `href` should be replaced with `[routerLink]`, otherwise this link may break if you deploy the UI on a subpath.

File: src/app/suggestion-notifications/qa/project-entry-import-modal/project-entry-import-modal.component.html
Comment (tdonohue) [inline src/app/suggestion-notifications/qa/project-entry-import-modal/project-entry-import-modal.component.html:21]: Again this should use `[routerLink]`

File: src/app/core/suggestion-notifications/qa/events/quality-assurance-event-data.service.spec.ts
Comment (tdonohue) [inline src/app/core/suggestion-notifications/qa/events/quality-assurance-event-data.service.spec.ts:1]: Starting from here, a lot of the code is under a folder named `suggestion-notifications`.  Since this is just called `Notifications` in the Admin menu, I wonder if we should consider renaming these directories to just be `notifications`? 

NOTE: If we feel this would be too difficult, I'm OK keeping this as-is.  It just seems odd to me to use the work "suggestions" as it may make this code more difficult to find for newer developers... they are more likely to search `notifications`

File: src/app/suggestion-notifications/suggestion-notifications.module.ts
Comment (tdonohue) [inline src/app/suggestion-notifications/suggestion-notifications.module.ts:1]: I see this is where we start having modules/services with the `suggestion-notifications` name.  I still find this name odd, but perhaps it'll make more sense for later PRs?

**Review by tdonohue — COMMENTED** [review]
@frabacche ; I gave this another look today.  My prior feedback has been resolved, but there are a few small suggestions I have for the new configurations you've added (see inline below).

I believe the **only remaining feedback** to address is to try to fix the small bugs mentioned in [my review of the backend](https://github.com/DSpace/DSpace/pull/8184#pullrequestreview-1756224759).  The following bugs still exist from that prior report:
> 1. Breadcrumbs are not working properly. They always only have one level. It should be possible to move from a Topic back to the list of topics using the breadcrumbs. (This is visible in the screenshot below where that page doesn't have a parent page in the breadcrumbs)
> 2. All small action buttons (Accept/Ignore/Reject) on a topic page have no accessibility text. This can be easily seen by performing a scan of the page using a browser accessibility plugin like ["axe DevTools"](https://chromewebstore.google.com/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd). See the screenshot.
> button-text-is-empty
> 3. I believe the same accessibility issue also exists with the "Actions" buttons on the "Quality Assurance" page.  The accessibility text on those buttons is just a number.
> 4. When I visit the page for a topic (like in the screenshot above), I see data logged to my Console window (in Chrome's DevTools).  Something looks to be sending all the JSON data to `console.log`? 
> ...
> 6. The ordering of the "Current Topics" on the initial "Quality Assurance" page (`/admin/notifications/quality-assurance/openaire`) is unclear & seems to randomly change at times.  _(WE AGREED it should be alphabetical)_
> ...
> 9. Finally, ... it'd be nice if there was a way to **disable** this feature altogether.  (By this, I mean that the OpenAIRE quality assurance page should not appear if OpenAIRE integration is somehow disabled. It could be via a boolean setting in `qaevents.cfg` or maybe it just checks to see if the `qaevents.openaire.broker-url` is specified. If so, it's enabled, if not it's disabled.)

All other feedback was addressed.  You mentioned the above are being "worked on", so this is just a reminder that I'm waiting on these before giving a +1 vote.  Overall, this is nearly ready. The sooner you can get to these changes, the sooner we can get this merged!

File: src/app/notifications/qa/events/quality-assurance-events.component.ts
Comment (paulo-graca) [inline src/app/notifications/qa/events/quality-assurance-events.component.ts:102]: not sure why openaire should be available here, shouldn't quality assurance be source agnostic?

File: src/app/core/notifications/qa/models/quality-assurance-event.model.ts
Comment (paulo-graca) [inline src/app/core/notifications/qa/models/quality-assurance-event.model.ts:25]: why openaire here?

File: src/app/core/notifications/qa/models/quality-assurance-event.model.ts
Comment (paulo-graca) [inline src/app/core/notifications/qa/models/quality-assurance-event.model.ts:74]: shouldn't be source agnostic?

**Review by paulo-graca — COMMENTED** [review]
Thank you, @LucaGiamminonni, for addressing my comments. I'm currently heavily involved in our local projects, and I don't have much time right now, and until the end of the year, to do thorough reviews. Please excuse me for that.

I just want to share with you some concerns I have and what I'm expecting the result to be for this PR:

I expect to have a consistent interface that doesn't break with what the user is used to. So, I expect the site navigation to be consistent.

I also expect the user to be content source-aware. It should be clear at any time that the suggested content came from a specific source (we can have others than OpenAIRE).

This is mainly an admin feature, so perhaps this comment doesn't make sense, but I didn't see themed components. Not sure if needed.

This feature should be source-agnostic, but I saw some references to OpenAIRE (properties, methods).

I'm not intending to block this PR. If @tdonohue understands that this PR can go as is, he can have my +1, and issues can be addressed later on.

File: src/app/notifications/qa/events/quality-assurance-events.component.ts
Comment (FrancescoMolinaro) [reply src/app/notifications/qa/events/quality-assurance-events.component.ts]: Hi @paulo-graca , I have adapted the config to use the source id to retrieve the url, so that we would need only a new config for future sources.

**Comment by tdonohue** [comment]
@frabacche : This appears to have 3 lint errors after the recent updates.  Run `yarn lint` locally to see them.  You may also be able to run `yarn lint --fix` to try to fix them automatically (doesn't always work).

**Review by tdonohue — APPROVED** [review]
👍 Thanks @frabacche , @FrancescoMolinaro and @LucaGiamminonni !  This all looks good to me now.  I gave it another test/review and verified my prior feedback has all been addressed.  I also checked @paulo-graca 's feedback and it also looks to be addressed, except for a request for making all components themeable. But, in my opinion, that can be addressed in a later PR as more features will need to be built on this initial PR.

**Comment by tdonohue** [comment]
Merging, as this has a +1 from myself, and a prior +1 from @paulo-graca .  I've also verified that all of @paulo-graca 's feedback appears to be addressed _except for the themed components_.  But those can be added in later PRs

Thanks again to @frabacche , @FrancescoMolinaro , @LucaGiamminonni and the entire 4Science team for their hard work on this new feature!

**Pull request merged**
